File: C/credit/credit.c
// From a credit card number find if the credit card is valid and if it's American Express, Visa or MasterCard
// Problem link: https://cs50.harvard.edu/x/2022/psets/1/credit/

#include <cs50.h>
#include <stdio.h>

int calculateChecksum(long number);
int calculateCardLength(long number);
int findStartingDigits(long number);

int main(void)
{
    // Propmt user for credit card number
    long cardNumber = get_long("Number: ");

    // Calculate checksum (Luhn’s Algorithm)
    int checksum = calculateChecksum(cardNumber);

    // Check Luhn's Algorithm
    if (checksum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Calculate card number length
    int length = calculateCardLength(cardNumber);

    // Find starting digits of card number
    int startingDigits = findStartingDigits(cardNumber);

    // Print type of card
    if (length == 15 && (startingDigits == 34 || startingDigits == 37))
    {
        printf("AMEX\n");
    }
    else if ((length == 13 || length == 16) && startingDigits / 10 == 4)
    {
        printf("VISA\n");
    }
    else if (length == 16 && startingDigits > 50 && startingDigits < 56)
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

int calculateChecksum(long number)
{
    long cardNumberCopy1 = number;
    long cardNumberCopy2 = number;
    int sum1 = 0; // sum of every other digit starting with the number’s second-to-last digit multiplied by 2
    int sum2 = 0; // sum of the digits that weren’t multiplied by 2 (the rest of the digits)

    cardNumberCopy1 /= 10; // remove last digit from card number
    sum2 = cardNumberCopy2 % 10; // add last digit to sum 2

    while (cardNumberCopy1 > 0)
    {
        int digit = (cardNumberCopy1 % 10) * 2; // get every other digit and multiply by 2
        // Add the products’ digits (not the products themselves) together
        if (digit < 10)
        {
            sum1 += digit;
        }
        else
        {
            int sumSingleDigits = (digit % 10) + (digit / 10); // get single digits and add them up
            sum1 += sumSingleDigits;
        }
        cardNumberCopy1 /= 100; // remove last two digits from card number

        cardNumberCopy2 /= 100; // remove last two digits from card number
        sum2 += cardNumberCopy2 % 10; // calculate sum of the digits that weren’t multiplied by 2
    }

    return sum1 + sum2;
}

int calculateCardLength(long number)
{
    int length = 0;
    while (number > 0)
    {
        number /= 10;
        length++;
    }
    return length;
}

int findStartingDigits(long number)
{
    while (number > 100)
    {
        number /= 10;
    }

    return number;
}

File: python/sentimental-credit/credit.py
# Implement a program that determines whether a provided credit card number is valid according to Luhn’s algorithm
# Problem link: https://cs50.harvard.edu/x/2022/psets/6/credit/

from cs50 import get_int
import sys
import re


# Prompt the user for credit card number
number = get_int("Number: ")

# Convert number to string
number_str = str(number)

# Length of number
length_number = len(number_str)

# Check credit card number length
if length_number != 15 and length_number != 13 and length_number != 16:
    print("INVALID\n")
    sys.exit()


# Apply Luhn’s algorithm
sum1 = 0
sum2 = 0

# Every other digit, starting with the number’s second-to-last digit, to multiply by 2 and add products’ digits together
for i in range(length_number - 2, -1, -2):
    n_multiplied = int(number_str[i]) * 2
    if n_multiplied > 9:
        for digit in str(n_multiplied):
            sum1 += int(digit)
    else:
        sum1 += n_multiplied

# Sum of digits that weren’t multiplied by 2 (starting from the end)
for i in range(length_number - 1, -1, -2):
    sum2 += int(number_str[i])

# Add both sums and check if last digit is 0
total_sum = str(sum1 + sum2)
if total_sum[len(total_sum) - 1] != "0":
    print("INVALID\n")
    sys.exit()


# Print type of card
if re.match("^4", number_str):
    print("VISA\n")
elif re.match("^34|37", number_str):
    print("AMEX\n")
elif re.match("^5[1-5]", number_str):
    print("MASTERCARD\n")
else:
    print("INVALID\n")